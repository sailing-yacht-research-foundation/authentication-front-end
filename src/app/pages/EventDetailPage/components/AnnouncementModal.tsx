import React from 'react';
import { useTranslation } from 'react-i18next';
import { toast } from 'react-toastify';
import { navigateToProfile, showToastMessageOnRequestError } from 'utils/helpers';
import { Modal, Spin, Form, Select, Radio } from 'antd';
import { useForm } from 'antd/lib/form/Form';
import { SyrfFieldLabel, SyrfFormButton, SyrfFormSelect, SyrfTextArea } from 'app/components/SyrfForm';
import { translations } from 'locales/translations';
import { Participant } from 'types/Participant';
import { getAcceptedAndSelfRegisteredParticipantByCalendarEventId } from 'services/live-data-server/participants';
import { CalendarEvent } from 'types/CalendarEvent';
import { sendMessageToVesselParticipants } from 'services/live-data-server/event-calendars';
import { ItemAvatar } from 'app/components/SyrfGeneral';
import { useHistory } from 'react-router-dom';
import { renderAvatar } from 'utils/user-utils';

const radioValue = {
    SEND_TO_ALL: 1,
    SEND_TO_SOME: 2,
}

interface IAnnouncementModal {
    event: Partial<CalendarEvent>,
    showModal: boolean,
    setShowModal: Function,
    reloadParent: Function
}

export const AnnouncementModal = ({ event, showModal, setShowModal, reloadParent }: IAnnouncementModal) => {

    const [isLoading, setIsLoading] = React.useState<boolean>(false);

    const [participants, setParticipants] = React.useState<Participant[]>([]);

    const [selectedRadioValue, setSelectedRadioValue] = React.useState<number>(radioValue.SEND_TO_ALL);

    const [form] = useForm();

    const { t } = useTranslation();

    const history = useHistory();

    const onFinish = async (values) => {

        const { message, participantIds } = values;

        setIsLoading(true);
        const response = await sendMessageToVesselParticipants(event.id!, {
            message: message,
            participantIds: participantIds ?? []
        });
        setIsLoading(false);

        if (response.success) {
            toast.success(t(translations.event_detail_page.successfully_delivered_your_message_to_the_competitors));
            hideModal();
            reloadParent();
        } else {
            showToastMessageOnRequestError(response.error);
        }
    }

    const hideModal = () => {
        setShowModal(false);
        form.resetFields();
    }

    const getAllEventParticipants = async () => {

        const response = await getAcceptedAndSelfRegisteredParticipantByCalendarEventId(event.id!, 1, 1000);

        if (response.success) {
            setParticipants(response.data?.rows);
        }
    }

    const renderParticipantsList = () => {
        return participants.map(item => <Select.Option style={{ padding: '5px' }} value={item.id}>
            <ItemAvatar onClick={(e) => navigateToProfile(e, item, history)} src={renderAvatar(item.profile?.avatar)} /> {item.publicName}
        </Select.Option>)
    }

    React.useEffect(() => {
        if (event.id)
            getAllEventParticipants();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [event]);

    const onRadioChanged = (e) => {
        setSelectedRadioValue(e.target.value);
    }

    const renderFormFieldsBaseOnRadioValue = () => {
        switch (selectedRadioValue) {
            case radioValue.SEND_TO_ALL:
                return <></>;
            case radioValue.SEND_TO_SOME:
                return <Form.Item
                    label={<SyrfFieldLabel>{t(translations.event_detail_page.select_competitors)}</SyrfFieldLabel>}
                    name="participantIds"
                    rules={[{ required: true, message: t(translations.forms.please_fill_out_this_field) }]}
                >
                    <SyrfFormSelect mode='multiple'
                        allowClear
                        maxTagCount={'responsive'}
                        filterOption={(input, option) => {
                            if (option) {
                                // the option.props.children[2] is where the name of the participant lies.
                                return option.props.children[2]?.toLowerCase().indexOf(input.toLowerCase()) >= 0
                            }

                            return false;
                        }}>
                        {renderParticipantsList()}
                    </SyrfFormSelect>
                </Form.Item>
        }
    }

    if (participants.length > 0)
        return (
            <Modal
                title={t(translations.event_detail_page.send_announcement)}
                bodyStyle={{ display: 'flex', justifyContent: 'center', overflow: 'hidden', flexDirection: 'column' }}
                visible={showModal}
                footer={null}
                onCancel={hideModal}
            >
                <Spin spinning={isLoading}>
                    <Form
                        form={form}
                        layout="vertical"
                        name="basic"
                        onFinish={onFinish}
                        style={{ width: '100%' }}
                    >
                        <Form.Item
                            label={<SyrfFieldLabel>{t(translations.event_detail_page.message)}</SyrfFieldLabel>}
                            name="message"
                            rules={[{ required: true, message: t(translations.forms.please_fill_out_this_field) }, {
                                max: 5000, message: t(translations.forms.please_input_no_more_than_characters, { numberOfChars: 5000 })
                            }]}
                        >
                            <SyrfTextArea />
                        </Form.Item>

                        <Form.Item
                            label={<SyrfFieldLabel>{t(translations.event_detail_page.participants)}</SyrfFieldLabel>}
                        >
                            <Radio.Group onChange={onRadioChanged} value={selectedRadioValue}>
                                <Radio value={radioValue.SEND_TO_ALL}>{t(translations.event_detail_page.send_to_all_competitors)}</Radio>
                                <Radio value={radioValue.SEND_TO_SOME}>{t(translations.event_detail_page.send_to_some_of_the_competitors)}</Radio>                        </Radio.Group>
                        </Form.Item>

                        {renderFormFieldsBaseOnRadioValue()}

                        <SyrfFormButton type="primary" htmlType="submit">
                            {t(translations.event_detail_page.send_announcement)}
                        </SyrfFormButton>
                    </Form>
                </Spin>
            </Modal>
        );

    return (
        <Modal
            title={t(translations.event_detail_page.send_announcement)}
            bodyStyle={{ display: 'flex', justifyContent: 'center', overflow: 'hidden', flexDirection: 'column' }}
            visible={showModal}
            cancelButtonProps={{ style: { display: 'none' } }}
            onOk={hideModal}
            onCancel={hideModal}
        >
            <span>{t(translations.event_detail_page.please_invite_at_least_1_competitor_to_send_announcement)}</span>
        </Modal>
    )
}