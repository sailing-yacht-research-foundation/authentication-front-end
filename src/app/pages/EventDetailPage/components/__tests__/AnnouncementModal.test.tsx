import * as React from 'react';
import { AnnouncementModal } from '../AnnouncementModal';
import * as AnnouncementModalModule from '../AnnouncementModal';
import { createRenderer } from 'react-test-renderer/shallow';
import { i18n } from 'locales/i18n';
import { translations } from 'locales/translations';
import { act, fireEvent, render, } from '@testing-library/react';

const shallowRenderer = createRenderer();

const modalShowComponent = (
    <AnnouncementModal event={{}} showModal={true} setShowModal={jest.fn} reloadParent={jest.fn} />
);

describe('AnnouncementModal', () => {
    it('should render and match snapshot', () => {
        shallowRenderer.render(modalShowComponent);
        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput).toMatchSnapshot();
    });

    // Use 'it' to test a single attribute of a target
    it('Hide if show modal flag is false', () => {

        shallowRenderer.render(<AnnouncementModal event={{}} showModal={false} setShowModal={jest.fn} reloadParent={jest.fn} />);

        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput.props.visible).not.toBeTruthy();
    });

    it('Show if show modal flag is true', async () => {
        shallowRenderer.render(modalShowComponent);

        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput.props.visible).toBeTruthy();
    });

    it('Modal title and content should match i18next translation language', async () => {
        const t = await i18n;

        shallowRenderer.render(modalShowComponent);

        expect(t(shallowRenderer.getRenderOutput().props.title)).toBe(t(translations.event_detail_page.send_announcement));
        expect(t(shallowRenderer.getRenderOutput().props.children.props.children)).toBe(t(translations.event_detail_page.please_invite_at_least_1_competitor_to_send_announcement));
    });

    it("It should have been rendered with correct received props", () => {
        const announcementModalSpy = jest.spyOn(AnnouncementModalModule, 'AnnouncementModal');

        render(<AnnouncementModal event={{}} showModal={true} setShowModal={jest.fn} reloadParent={jest.fn} />);

        expect(announcementModalSpy).toHaveBeenCalledWith({
            event: {},
            showModal: true,
            setShowModal: jest.fn,
            reloadParent: jest.fn,
        }, {});
    });

    it("It should call setShowModal when pressing cancel text", () => {
        const setShowModalMock = jest.fn();

        const { getByText } = render(<AnnouncementModal event={{}} showModal={true} setShowModal={setShowModalMock} reloadParent={jest.fn} />);

        act(() => {
            fireEvent.click(getByText('Cancel'));
        });

        expect(setShowModalMock).toHaveBeenCalledWith(false);
    });
});