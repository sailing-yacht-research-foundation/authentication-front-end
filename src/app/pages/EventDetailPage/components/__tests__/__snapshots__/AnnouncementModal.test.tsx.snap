// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnouncementModal should render and match snapshot 1`] = `
<Modal
  bodyStyle={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
      "overflow": "hidden",
    }
  }
  cancelButtonProps={
    Object {
      "style": Object {
        "display": "none",
      },
    }
  }
  confirmLoading={false}
  okType="primary"
  onCancel={[Function]}
  onOk={[Function]}
  title="Send Announcement"
  visible={true}
  width={520}
>
  <span>
    Please invite at least 1 competitor to send announcement.
  </span>
</Modal>
`;
